File: python/chdir.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 21 17:18:15 2015

chdir script customized for the LINCS L1000 pipeline.

Updated on Dec. 2 2015

@author: qn
"""

import numpy as np

def chdir(A, B, r=1):
#  calculate a characteristic direction for a gene expression dataset 
#  A: control numeric gene expressoion data
#  B: experiment numeric gene expression data
#  b: return value, a vector of n-components, representing the characteristic
#     direction of that gene expression dataset. n equals to the number of genes 
#     in the expression dataset.
#  r: regulaized term. A parameter that smooths the covariance matrix and reduces
#     potential noise in the dataset.

# * CHANGED *
# 1. All validation happens in a common function that validates data for all
#    differential expression methods.
# 2. Use numpy to concatenate A and B.


    ctrlCount = np.shape(A)[1]
    expmCount = np.shape(B)[1]
# place control gene expression data and experiment gene expression data into
# one matrix X.
    X = np.concatenate((A,B), axis = 1).T

# get the number of samples (rowCount), namely, the total number of replicates in   
# control and experiment. Also get the number of genes (colCount)
    (rowCount,colCount) = np.shape(X)

#  the number of output components desired from PCA. We only want to calculate
#  the chdir in a subspace that capture most variance in order to save computation 
#  workload. The number is set 20 because considering the number of genes usually 
#  present in an expression matrix 20 components would  capture most of the variance.
    if 30 > rowCount-1:
        maxComponentsNum = rowCount - 1
    else:
        maxComponentsNum = 30

# use the nipals PCA algorithm to calculate scores, loadings, and explained_var. 
# explained_var are the variances captured by each component 
    scores, loadings, explained_var = nipals(X,maxComponentsNum,1e5,1e-4)
    scores = scores.T
    loadings = loadings.T

# We only want components that cpature 95% of the total variance or a little above.
    captured_variance = 0
    for i in range(len(explained_var)):
        captured_variance += explained_var[i]
        if captured_variance > 0.999:
            break

# slice scores and loadings to only that number of components.
    scores = scores[0:i+1] # R in Neil's algorithm
    loadings = loadings[0:i+1] # V in Neil's algorithm

    scores = scores.T
    loadings = loadings.T

# the difference between experiment mean vector and control mean vector.
    meanvec = np.mean(B,axis=1,keepdims=True) - np.mean(A,axis=1,keepdims=True)

# All the following steps calculate shrunkMats. Refer to Neil's paper for detail.
# ShrunkenMats are the covariance matrix that is placed as denominator 
# in LDA formula. Notice the shrunkMats here is in the subspace of those components
# that capture about 95% of total variance.
    ctrlScores = scores[0:ctrlCount,:]
    expmScores = scores[(ctrlCount):(ctrlCount+expmCount),:]
    Dd = (np.dot(ctrlScores.T,ctrlScores)+np.dot(expmScores.T,expmScores))/(ctrlCount+expmCount-2)
#    Dd = np.diag(np.diag(Dd))
    sigma = np.mean(np.diag(Dd))
    shrunkMats = np.dot(r,Dd)+ sigma*(1-r)*np.eye(np.shape(Dd)[0])
    invMat = np.linalg.inv(shrunkMats)

# The LDA formula.
# np.dot(np.dot(loadings,shrunkMats),loadings.T) transforms the covariance 
# matrix from the subspace to full space.
    b = np.dot(loadings,np.dot(invMat,np.dot(loadings.T,meanvec)))

# normlize b to unit vector
    b /= np.linalg.norm(b)

# ! CHANGED !
# Do not sort the genes before returning them. This is because Neil's unit
# test data is not sorted.
# 
# ! CHANGED!
# Return values as Python list for consistent user interface.
    print('Done chdir')
    return b



def nipals(X,a,it=100,tol=1e-4):
    # Nipals algorithm for Principal Component Analysis
    # This function is written largely based on nipals function from R chemometrics package.

    X = np.array(X)
    (obsCount,varCount) = np.shape(X)
    Xh = X - np.tile(np.mean(X,axis=0),(obsCount,1))

    T = np.zeros((obsCount,a))
    P = np.zeros((varCount,a))
    pcvar = np.zeros(varCount)
    varTotal = np.sum(np.var(Xh,axis=0))
    currVar = varTotal
    nr = 0

    for h in range(a):
        th = np.reshape(Xh[:,0],(obsCount,-1))
        ende = False

        while not ende:
            nr = nr + 1
            ph = np.dot(Xh.T,th)/np.dot(th.T,th)
            ph = ph/np.sqrt(np.dot(ph.T,ph))
            thnew = np.dot(Xh,ph)/np.dot(ph.T,ph)
            prec = np.dot((thnew-th).T,(thnew-th))
            th = thnew

            if prec <= np.power(tol,2):
                ende = True
            if it <= nr:
                ende = True
                print('Iteration stops without convergence')

        Xh = Xh - np.dot(th,ph.T)
        T[:,h] = th[:,0]
        P[:,h] = ph[:,0]
        oldVar = currVar
        currVar = np.sum(np.var(Xh,axis=0))
        pcvar[h] = ( oldVar - currVar )/varTotal
        nr = 0

    return T, P, pcvar

File: R/chdir.R
"chdir" <-
function(ctrl,expm,genes,r=1)
# This function caclulates the characteristic direction for a gene expression dataset.
#  	ctrl: control gene expressoion data, a matrix object
#  	expm: experiment gene expression data, a matrix object
#  	b: return value, a vector of n-components, representing the characteristic
#          direction of the gene expression dataset. n equals to the number of genes in the 
#          expression dataset. b is also a matrix object. b is sorted by its components' 
#          absolute values in descending order.
#  	r: regularized term. A parameter that smooths the covariance matrix and reduces
#          potential noise in the dataset. The default value for r is 1, no regularization.
#
#       For the input matrix rows are genes and columns are gene expression profiles.
#       r is the regulization term ranging [0,1]. b is the characteristic direction.
#       ctrl(control) and expm(experiment) matrices should have the same number
#       of genes(rows). 
#
#       Author: Qiaonan Duan
#       Ma'ayan Lab, Icahn School of Medicine at Mount Sinai
#       Jan.13, 2014
#
#		Add gene symbols to results. Apr. 4, 2014

{

     if(dim(ctrl)[1]!=dim(expm)[1]){
	stop('Control expression data must have equal number of genes as experiment expression data!')
}

     if(any(is.na(ctrl))||any(is.na(expm))){
	stop('Control expression data and experiment expression data have to be real numbers. NA was found!')
}


# There should be variance in expression values of each gene. If  
# gene expression values of a gene are constant, it would dramatically
# affect the LDA caculation and results in a wrong answer.
constantThreshold <- 1e-5;
ctrlConstantGenes <- diag(var(t(ctrl))) < constantThreshold
expmConstantGenes <- diag(var(t(expm))) < constantThreshold

if (any(ctrlConstantGenes)){
	errMes <- sprintf('%s row(s) in control expression data are constant. Consider Removing the row(s).',paste(as.character(which(ctrlConstantGenes)),collapse=','))
	stop(errMes)
}else if(any(expmConstantGenes)){
	errMes <- sprintf('%s row(s) in experiment expression data are constant. Consider Removing the row(s).',paste(as.character(which(expmConstantGenes)),collapse=','))
	stop(errMes)
}

# place control gene expression data and experiment gene expression data into
# one matrix
combinedData <- cbind(ctrl,expm)

# get the number of samples, namely, the total number of replicates in  control 
# and experiment. 
dims <- dim(combinedData)
samplesCount <- dims[2]

# the number of output components desired from PCA. We only want to calculate
# the chdir in a subspace that capture most variance in order to save computation 
# workload. The number is set 20 because considering the number of genes usually 
# present in an expression matrix 20 components would capture most of the variance.
componentsCount <- min(c(samplesCount-1,20))


# use the nipals PCA algorithm to calculate R, V, and pcvars. nipals algorithm
# has better performance than the algorithm used by R's builtin PCA function.
# R are scores and V are coefficients or loadings. pcvars are the variances 
# captured by each component 
pcaRes <- nipals(t(combinedData),componentsCount,1e5,1e-4)
R <- pcaRes$T
V <- pcaRes$P
pcvars <- pcaRes$pcvar


# we only want components that cpature 95% of the total variance or a little above.
# cutIdx is the index of the compoenent, within which the variance is just equal
# to or a little greater than 95% of the total.
cutIdx <- which(cumsum(pcvars)>0.95)
if(length(cutIdx)==0){
	cutIdx <- componentsCount
}else{
	cutIdx <- cutIdx[1]
}

# slice R and V to only that number of components.
R <- R[,1:cutIdx]
V <- V[,1:cutIdx]

# the difference between experiment mean and control mean.
meanvec <- rowMeans(expm) - rowMeans(ctrl)


# all the following steps calculate shrunkMats. Refer to the ChrDir paper for detail.
# ShrunkenMats are the covariance matrix that is placed as denominator 
# in LDA formula. Notice the shrunkMats here is in the subspace of those components
# that capture about 95% of total variance.
Dd <- t(R)%*%R/samplesCount
Dd <- diag(diag(Dd))
sigma <- mean(diag(Dd))
shrunkMats <- r*Dd + sigma*(1-r)*diag(dim(R)[2])

# The LDA formula.
#  V%*%solve(shrunkMats)%*%t(V) transforms the covariance matrix from the subspace to full space.
b <- V%*%solve(shrunkMats)%*%t(V)%*%meanvec

# normlize b to unit vector
b <- b*as.vector(sqrt(1/t(b)%*%b))

# sort b to by its components' absolute value in decreasing order and get the 
# sort index
sortRes <- sort(abs(b),decreasing=TRUE,index.return=TRUE)

# sort b by the sort index
bSorted <- as.matrix(b[sortRes$ix])
# sort genes by the sort index
genesSorted <- genes[sortRes$ix]
# assign genesSorted as the row names of bSorted
rownames(bSorted) <- genesSorted

# return bSorted
bSorted <- bSorted
}

